File: Eikonal_solver/Python/fast_march.py
import numpy as np
from initialize import initialize_4D
from heapq import heapify, heappush, heappop
from Grid import Grid

def fast_march(grid, xmin, xmax, ymin, ymax, zmin, zmax, qmin, qmax, boundary, F):
    # Initialize
    grid, narrow_band, not_known = initialize_4D(grid, xmin, xmax,
                                                       ymin, ymax,
                                                       zmin, zmax,
                                                       qmin, qmax,
                                                       boundary)
    print "Done with initialize"

    # March forward
    grid = march_forward(grid, xmin, xmax, ymin, ymax, zmin, zmax, qmin, qmax,
                         narrow_band, not_known, F)
    print "Done with fast march"

    return grid

def march_forward(grid, xmin, xmax, ymin, ymax, zmin, zmax, qmin, qmax,
                  narrow_band, not_known, F):
    n1, n2, n3, n4 = grid.shape
    dx = (xmax - xmin) / float(n1 - 1)
    dy = (ymax - ymin) / float(n2 - 1)
    dz = (zmax - zmin) / float(n3 - 1)
    dq = (qmax - qmin) / float(n4 - 1)
    heapify(narrow_band)

    while not_known > 0:
        # step 1
        A = heappop(narrow_band)

        # step 2
        i, j, k, l = A[1], A[2], A[3], A[4]
        grid[i, j, k, l] = -A[0]

        # step 3
        gip1 = grid[i+1, j, k, l]
        # If Trial or Far, update value
        if gip1 > 0:
            grid[i+1, j, k, l] = advance(grid, i+1, j, k, l, dx, dy, dz, dq,
                                      F, xmin, ymin, zmin, qmin)
            # If far, add to narrow band
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i+1, j, k, l], i+1, j, k, l))

        gip1 = grid[i-1, j, k, l]
        if gip1 > 0:
            grid[i-1, j, k, l] = advance(grid, i-1, j, k, l, dx, dy, dz, dq,
                                      F, xmin, ymin, zmin, qmin)
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i-1, j, k, l], i-1, j, k, l))

        gip1 = grid[i, j+1, k, l]
        if gip1 > 0:
            grid[i, j+1, k, l] = advance(grid, i, j+1, k, l, dx, dy, dz, dq,
                                      F, xmin, ymin, zmin, qmin)
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i, j+1, k, l], i, j+1, k, l))

        gip1 = grid[i, j-1, k, l]
        if gip1 > 0:
            grid[i, j-1, k, l] = advance(grid, i, j-1, k, l, dx, dy, dz, dq,
                                      F, xmin, ymin, zmin, qmin)
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i, j-1, k, l], i, j-1, k, l))

        gip1 = grid[i, j, k+1, l]
        if gip1 > 0:
            grid[i, j, k+1, l] = advance(grid, i, j, k+1, l, dx, dy, dz, dq,
                                      F, xmin, ymin, zmin, qmin)
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i, j, k+1, l], i, j, k+1, l))

        gip1 = grid[i, j, k-1, l]
        if gip1 > 0:
            grid[i, j, k-1, l] = advance(grid, i, j, k-1, l, dx, dy, dz, dq,
                                      F, xmin, ymin, zmin, qmin)
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i, j, k-1, l], i, j, k-1, l))

        gip1 = grid[i, j, k, l+1]
        if gip1 > 0:
            grid[i, j, k, l+1] = advance(grid, i, j, k, l+1, dx, dy, dz, dq,
                                         F, xmin, ymin, zmin, qmin)
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i, j, k, l+1], i, j, k, l+1))

        gip1 = grid[i, j, k, l-1]
        if gip1 > 0:
            grid[i, j, k, l-1] = advance(grid, i, j, k, l-1, dx, dy, dz, dq,
                                         F, xmin, ymin, zmin, qmin)
            if gip1 == np.inf:
                heappush(narrow_band, (grid[i, j, k, l-1], i, j, k, l-1))
        not_known -= 1

    return grid

def advance(grid, i, j, k, l, dx, dy, dz, dq, F, xmin, ymin, zmin, qmin):
    F_ij = F(i*dx + xmin, j*dy + ymin, k*dz + zmin, l*dq + qmin)
    T_H = min(abs(grid[i+1, j, k, l]), abs(grid[i-1, j, k, l]))
    T_V = min(abs(grid[i, j+1, k, l]), abs(grid[i, j-1, k, l]))
    T_Z = min(abs(grid[i, j, k+1, l]), abs(grid[i, j, k-1, l]))
    T_Q = min(abs(grid[i, j, k, l+1]), abs(grid[i, j, k, l-1]))
    T = [T_H, T_V, T_Z, T_Q]
    T = [T_j for T_j in T if T_j != np.inf]
    T.sort()
    j = len(T)
    while sum([(T[j-1] - T[i])**2 for i in range(j-1)]) >= dx**2 / float(F_ij**2):
        j = j - 1
    return 1./j*sum([T[i] for i in range(j)])\
         + 1./j*np.sqrt((sum([T[i] for i in range(j)]))**2
         - j*(sum([T[i]**2 for i in range(j)]) - dx**2 / float(F_ij**2)))

def exact_square(n, xmin, xmax, r):
    grid = -np.zeros((n, n, n, n))
    h = (xmax - xmin)/float(n-1)
    for i in range(n):
        for j in range(n):
            for k in range(n):
                for l in range(n):
                    x = abs(i*h + xmin)
                    y = abs(j*h + xmin)
                    z = abs(k*h + xmin)
                    q = abs(l*h + xmin)
                    if x > 0.9 or y > 0.9 or z > 0.9 or q > 0.9:
                        grid[i,j,k,l] = -0
                    else:
                        grid[i,j,k,l] = -(0.9 - max(x, y, z, q))
    return grid

def exact_circle(n, xmin, xmax, r):
    grid = -np.zeros((n, n, n, n))
    h = (xmax - xmin)/float(n-1)
    for i in range(n):
        for j in range(n):
            for k in range(n):
                for l in range(n):
                    x = abs(i*h + xmin)
                    y = abs(j*h + xmin)
                    z = abs(k*h + xmin)
                    q = abs(l*h + xmin)
                    if x**2 + y**2 + z**2 + q**2 > r**2:
                        grid[i,j,k,l] = -0
                    else:
                        grid[i,j,k,l] = -(0.9 - np.sqrt(x**2 + y**2 + z**2 + q**2))
    return grid

def compute_dist_dU(C=1, m=10):
    # boundary used in the thesis
    n1, n2, n3, n4 = [m]*4
    grid = -np.zeros((n1, n2, n3, n4))
    xmin, xmax = -np.sqrt(C), np.sqrt(C)
    ymin, ymax = -np.sqrt(C), np.sqrt(C)
    zmin, zmax = -np.sqrt(C), np.sqrt(C)
    qmin, qmax = -np.sqrt(C), np.sqrt(C)
    def boundary(x, y, z, q):
        # compute both eigenvalues
        lambda_1 = 1/2. * (-(x**2 + y**2) + C) + 1/2. \
                  * np.sqrt(x**4 + 2*x**2*y**2 - 4*x**2*z - 8*x*y*q
                            + y**4 + 4*y**2*z + 4*z**2 + 4*q**2)
        lambda_2 = 1/2. * (-(x**2 + y**2) + C) - 1/2. \
                  * np.sqrt(x**4 + 2*x**2*y**2 - 4*x**2*z - 8*x*y*q
                            + y**4 + 4*y**2*z + 4*z**2 + 4*q**2)
        if lambda_1 > 0 and lambda_2 > 0:
            return -lambda_1*lambda_2
        return 0
    F = lambda x, y, z, q: 1
    grid = fast_march(grid, xmin, xmax, ymin, ymax, zmin, zmax, qmin, qmax,
                      boundary, F)
    return grid

def write_to_file(grid, xmin, xmax, filename="python_dist.txt"):
    n = grid.shape[0]
    with open(filename, "w") as outfile:
        outfile.write("%g %g %g \n" % (n, xmin, xmax))
        for point in grid.elements:
            outfile.write("%03d%03d%03d%03d " % (point))
            outfile.write("%g\n" % grid[point])
    return


if __name__ == '__main__':
    solution_grid = Grid(standard_value=0, shape=[30] * 4)

    def boundary_function(x, y, z, q):
        return x ** 2 + y ** 2 + z ** 2 + q ** 2 - 1

    fast_march(grid=solution_grid,
               xmin=-2, xmax=2,
               ymin=-2, ymax=2,
               zmin=-2, zmax=2,
               qmin=-2, qmax=2,
               boundary=boundary_function,
               F=lambda x, y, z, q: 1)
    print solution_grid[15, 15, 15, 15]

File: Eikonal_solver/C/fast_march.cpp
#include <iostream>
#include <algorithm> // for heap
#include "Grid.h"
#include "initialize.h"
#include "fast_march.h"
using namespace std;

void fast_march(Grid *grid, double xmin, double xmax,
                            double ymin, double ymax,
                            double zmin, double zmax,
                            double qmin, double qmax,
                            double boundary(double, double, double, double)) {
    // initialize
    Initialized init = initialize(grid, xmin, xmax,
                                        ymin, ymax,
                                        zmin, zmax,
                                        qmin, qmax,
                                        boundary);
    cout << "Done with initialize" << endl;

    // March forward
    march_forward(grid, xmin, xmax,
                        ymin, ymax,
                        zmin, zmax,
                        qmin, qmax,
                        boundary,
                        init.narrowband,
                        init.not_known);
    cout << "Done with fast march" << endl;
}

void march_forward(Grid *grid, double xmin, double xmax,
                               double ymin, double ymax,
                               double zmin, double zmax,
                               double qmin, double qmax,
                               double boundary(double, double, double, double),
                               std::vector<ValueIndex> narrowband,
                               long not_known) {
    int n1 = grid->getShape().a;
    int n2 = grid->getShape().b;
    int n3 = grid->getShape().c;
    int n4 = grid->getShape().d;

    double dx = (xmax - xmin) / (n1 - 1);
    double dy = (ymax - ymin) / (n2 - 1);
    double dz = (zmax - zmin) / (n3 - 1);
    double dq = (qmax - qmin) / (n4 - 1);

    make_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);

    ValueIndex A;
    int i; int j; int k; int l;
    int index[4];
    double gip1;
    double value;
    while(not_known > 0) {
        // Step 1
        A = narrowband.front();
        pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
        // remove A from vector
        narrowband.pop_back();

        // Step 2
        i = A.i; j = A.j;
        k = A.k; l = A.l;
        grid->setItem((int[4]) {i,j,k,l}, -A.value);

        // Step 3
        gip1 = grid->getItem((int[4]) {i+1, j, k, l});
        // If Trial of Far, update value
        if(gip1 > 0) {
            value = advance(grid, i+1, j, k, l, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            grid->setItem((int[4]) {i+1, j, k, l}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i+1, j, k, l};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        gip1 = grid->getItem((int[4]) {i-1, j, k, l});
        if(gip1 > 0) {
            value = advance(grid, i-1, j, k, l, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            grid->setItem((int[4]) {i-1, j, k, l}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i-1, j, k, l};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        gip1 = grid->getItem((int[4]) {i, j+1, k, l});
        if(gip1 > 0) {
            value = advance(grid, i, j+1, k, l, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            grid->setItem((int[4]) {i, j+1, k, l}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i, j+1, k, l};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        gip1 = grid->getItem((int[4]) {i, j-1, k, l});
        if(gip1 > 0) {
            value = advance(grid, i, j-1, k, l, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            grid->setItem((int[4]) {i, j-1, k, l}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i, j-1, k, l};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        gip1 = grid->getItem((int[4]) {i, j, k+1, l});
        if(gip1 > 0) {
            value = advance(grid, i, j, k+1, l, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            //grid->setItem(index, value);
            grid->setItem((int[4]) {i, j, k+1, l}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i, j, k+1, l};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        gip1 = grid->getItem((int[4]) {i, j, k-1, l});
        if(gip1 > 0) {
            value = advance(grid, i, j, k-1, l, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            //grid->setItem(index, value);
            grid->setItem((int[4]) {i, j, k-1, l}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i, j, k-1, l};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        gip1 = grid->getItem((int[4]) {i, j, k, l+1});
        if(gip1 > 0) {
            value = advance(grid, i, j, k, l+1, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            //grid->setItem(index, value);
            grid->setItem((int[4]) {i, j, k, l+1}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i, j, k, l+1};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        gip1 = grid->getItem((int[4]) {i, j, k, l-1});
        if(gip1 > 0) {
            value = advance(grid, i, j, k, l-1, dx, dy, dz, dq,
                            xmin, ymin, zmin, qmin);
            grid->setItem((int[4]) {i, j, k, l-1}, value);
            // If Far, add to narrow band
            if(gip1 == INFINITY) {
                ValueIndex v = {value, i, j, k, l-1};
                // add to vector
                narrowband.push_back(v);
                // push to heap
                pop_heap (narrowband.begin(), narrowband.end(), compare_ValueIndex);
            }
        }

        not_known--;
    }

}

bool compare_ValueIndex(ValueIndex v1, ValueIndex v2) {
    return v1.value > v2.value;
}

double advance(Grid *grid, int i, int j, int k, int l,
               double dx, double dy, double dz, double dq,
               double xmin, double ymin, double zmin, double qmin) {
    std::vector<double> T;
    int index1[4] = {i+1, j, k, l}; int index2[4] = {i-1, j, k, l};
    double T_H = min(abs(grid->getItem(index1)), abs(grid->getItem(index2)));
    if(T_H != INFINITY) {
        T.push_back(T_H);
    }
    index1[0] = i; index1[1] = j+1; index2[0] = i; index2[1] = j-1;
    double T_V = min(abs(grid->getItem(index1)), abs(grid->getItem(index2)));
    if(T_V != INFINITY) {
        T.push_back(T_V);
    }
    index1[1] = j; index1[2] = k+1; index2[1] = j; index2[2] = k-1;
    double T_Z = min(abs(grid->getItem(index1)), abs(grid->getItem(index2)));
    if(T_Z != INFINITY) {
        T.push_back(T_Z);
    }
    index1[2] = k; index1[3] = l+1; index2[2] = k; index2[3] = l-1;
    double T_Q = min(abs(grid->getItem(index1)), abs(grid->getItem(index2)));
    if(T_Q != INFINITY) {
        T.push_back(T_Q);
    }

    std::sort (T.begin(), T.end());
    // NB: n here is j in python code
    int n = T.size();
    double dx_square = pow(dx, 2);
    while(sum_squared(T, n) >= dx_square) {
        n--;
    }
    double v1 = 0;
    double v2 = 0;
    for(int i = 0; i < n; i++) {
        v1 += T[i];
        v2 += pow(T[i], 2);
    }
    double new_value = v1 + sqrt(pow(v1, 2) - n*(v2 - dx_square));
    new_value *= 1./n;
    return new_value;
}

double sum_squared(std::vector<double> T, int j) {
    double sum = 0;
    for(int i = 0; i < j-1; i++) {
        sum += pow(T[j-1] - T[i], 2);
    }
    return sum;
}

